utils/dataset_manifest/utils.py

```diff
@@ -21,4 +24,160 @@ def rotate_image(image, angle):
 def md5_hash(frame):
     if isinstance(frame, VideoFrame):
         frame = frame.to_image()
-    return hashlib.md5(frame.tobytes()).hexdigest() # nosec
\ No newline at end of file
+    return hashlib.md5(frame.tobytes()).hexdigest() # nosec
+
+def _define_data_type(media):
+    media_type, _ = mimetypes.guess_type(media)
+    if media_type:
+        return media_type.split('/')[0]
+
+def is_video(media_file):
+    return _define_data_type(media_file) == 'video'
+
+def is_image(media_file):
+    return _define_data_type(media_file) == 'image'
+
+def _list_and_join(root):
+    files = os.listdir(root)
+    for f in files:
+        yield os.path.join(root, f)
+
+def _prepare_context_list(files, base_dir):
+    return sorted(map(lambda x: os.path.relpath(x, base_dir), filter(is_image, files)))
+
+# Expected 2D format is:
+# data/
+#   00001.png
+#   related_images/
+#     00001_png/
+#       context_image_1.jpeg
+#       context_image_2.png
+def detect_related_images_2D(image_paths, root_path):
+    related_images = {}
+    latest_dirname = ''
+    related_images_exist = False
+
+    for image_path in sorted(image_paths):
+        rel_image_path = os.path.relpath(image_path, root_path)
+        dirname = os.path.dirname(image_path)
+        related_images_dirname = os.path.join(dirname, 'related_images')
+        related_images[rel_image_path] = []
+
+        if latest_dirname == dirname and not related_images_exist:
+            continue
```

Comment:
@nmanovic 

This is updated implementation for 2D. We do not find context images for images where parent directory does not include "related_images". Something like this (and some other optimizations) work for 3D case.